alacritty/src/input/keyboard.rs

```diff
@@ -0,0 +1,592 @@
+    let kitty_event_type = mode.contains(TermMode::REPORT_EVENT_TYPES)
+        && (key.repeat || key.state == ElementState::Released);
+
+    let context = KeyEscSequenceBuildingContext {
+        mode,
+        modifiers,
+        kitty_seq,
+        kitty_encode_all,
+        kitty_event_type,
+    };
+
+    let sequence_base = context
+        .prepare_numpad_base(&key)
+        .or_else(|| context.prepare_named_base(&key))
+        .or_else(|| context.prepare_text_base(&key));
+
+    let Some(KeyEscSequenceBase { payload, terminator }) = sequence_base else {
+        return Vec::new();
+    };
+
+    let mut payload = format!("\x1b[{}", payload);
+
+    // Add modifiers information.
+    if kitty_event_type
+        || !modifiers.is_empty()
+        || (mode.contains(TermMode::REPORT_ASSOCIATED_TEXT) && key.text.is_some())
+    {
+        payload.push_str(&format!(";{}", modifiers.esc_sequence_encoded()));
+    }
+
+    // Push event types.
+    if kitty_event_type {
+        payload.push(':');
+        let event_type = match key.state {
+            _ if key.repeat => '2',
+            ElementState::Pressed => '1',
+            ElementState::Released => '3',
+        };
+        payload.push(event_type);
+    }
+
+    // Associated text is not reported for the control/alt/logo key presses.
+    if mode.contains(TermMode::REPORT_ASSOCIATED_TEXT)
+        && key.state != ElementState::Released
+        && (modifiers.is_empty() || modifiers == EscSequenceModifiers::SHIFT)
+    {
+        if let Some(text) = key.text {
+            let mut codepoints = text.chars().map(u32::from);
+            if let Some(codepoint) = codepoints.next() {
+                payload.push_str(&format!(";{codepoint}"));
+            }
+            // Push the rest of the chars.
+            for codepoint in codepoints {
+                payload.push_str(&format!(":{codepoint}"));
+            }
+        }
+    }
+
+    payload.push(terminator);
+
+    payload.into_bytes()
+}
+
+/// Helper to create esc sequence payloads from [`KeyEvent`].
+pub struct KeyEscSequenceBuildingContext {
+    mode: TermMode,
+    /// The emitted sequence should follow the kitty keyboard protocol.
+    kitty_seq: bool,
+    /// Encode all the keys according to the protocol.
+    kitty_encode_all: bool,
+    /// Report event types.
+    kitty_event_type: bool,
+    modifiers: EscSequenceModifiers,
+}
+
+impl KeyEscSequenceBuildingContext {
+    /// Prepare key with the text attached to it.
+    fn prepare_text_base(&self, key: &KeyEvent) -> Option<KeyEscSequenceBase> {
+        match key.logical_key.as_ref() {
+            Key::Character(character) => {
+                // We could have compose sequience, which has more than one character. Thus
+                // ignore it when we have to encode all.
+                if self.kitty_encode_all && character.chars().count() > 1 {
+                    Some(KeyEscSequenceBase::new("0".into(), 'u'))
```

Comment:
Why are we passing 0 as payload here? Why is the character just 'u'?

Reply:
Hm, I read it wrongly initially. Basically we should send `0` only when assoc text is requested and when report_all is also set.

See https://sw.kovidgoyal.net/kitty/keyboard-protocol/#text-as-code-points .